tools/testtool/main.go

```diff
@@ -0,0 +1,203 @@
+	f.StringVar(&tc.KernelContainer, "kernel-container", tc.KernelContainer, "Container to use for kernel files")
+	f.StringVar(&tc.KernelPath, "kernel-path", tc.KernelPath, "Path where to find the kernel image")
+	f.StringVar(&tc.QEMUContainer, "qemu-container", tc.QEMUContainer, "Container to use for QEMU files")
+	f.StringVar(&tc.QEMUCmd, "qemu-cmd", tc.QEMUCmd, "QEMU command with platform specific flags")
+	f.StringVar(&tc.QEMUPath, "qemu-path", tc.QEMUPath, "Path where to find the QEMU binary")
+	f.StringVar(&tc.BIOSPath, "bios-path", tc.BIOSPath, "Path where to find the BIOS image")
+}
+
+var configs = map[string]TestEnvConfig{
+	"amd64": {
+		KernelContainer: "ghcr.io/hugelgupf/vmtest/kernel-amd64:main",
+		KernelPath:      "/bzImage",
+		QEMUContainer:   "ghcr.io/hugelgupf/vmtest/qemu:main",
+		QEMUCmd:         "qemu-system-x86_64 -L %s -m 1G",
+		QEMUPath:        "/zqemu/bin/qemu-system-x86_64",
+		BIOSPath:        "/zqemu/pc-bios",
+	},
+	"arm": {
+		KernelContainer: "ghcr.io/hugelgupf/vmtest/kernel-arm:main",
+		KernelPath:      "/zImage",
+		QEMUContainer:   "ghcr.io/hugelgupf/vmtest/qemu:main",
+		QEMUCmd:         "qemu-system-arm -M virt,highmem=off -L %s",
+		QEMUPath:        "/zqemu/bin/qemu-system-arm",
+		BIOSPath:        "/zqemu/pc-bios",
+	},
+	"arm64": {
+		KernelContainer: "ghcr.io/hugelgupf/vmtest/kernel-arm64:main",
+		KernelPath:      "/Image",
+		QEMUContainer:   "ghcr.io/hugelgupf/vmtest/qemu:main",
+		QEMUCmd:         "qemu-system-aarch64 -machine virt -cpu max -m 1G -L %s",
+		QEMUPath:        "/zqemu/bin/qemu-system-arm64",
+		BIOSPath:        "/zqemu/pc-bios",
+	},
+}
+
+func defaultConfig() TestEnvConfig {
+	arch := os.Getenv("UROOT_TESTARCH")
+	if c, ok := configs[arch]; ok {
+		return c
+	}
+	if c, ok := configs[runtime.GOARCH]; ok {
+		return c
+	}
+	// On other architectures, user has to provide all values via flags.
+	return TestEnvConfig{}
+}
+
+func run() error {
+	config := defaultConfig()
+	config.RegisterFlags(flag.CommandLine)
+	containerized := flag.Bool("containerized", false, "Run the provided command in a containerized environment instead of your native environment")
+	flag.Parse()
+	if flag.NArg() < 2 {
+		return fmt.Errorf("too few arguments")
+	}
+
+	ctx := context.Background()
+	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
+	if err != nil {
+		return fmt.Errorf("unable to connect to client: %w", err)
+	}
+	defer client.Close()
+
+	artifacts := client.
+		Container().
+		From(config.QEMUContainer).
+		WithFile(config.KernelPath, client.Container().From(config.KernelContainer).File(config.KernelPath)).
+		Directory("/")
+
+	if *containerized {
+		return runContainerized(ctx, client, artifacts, config.KernelPath, config.QEMUCmd, flag.Args())
+	}
+
+	return runNatively(ctx, artifacts, config.KernelPath, config.QEMUCmd, flag.Args())
+}
+
+func runNatively(
+	ctx context.Context,
+	artifacts *dagger.Directory,
+	kpath, qemuCmd string,
+	args []string,
+) error {
```

Comment:
preferred inline due to https://google.github.io/styleguide/go/decisions#function-formatting

Reply:
yep

Reply:
Done